### ISSUE TITLE
exists() of tool in use is never called

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#3005]
This issue was originally created at: 2015-04-21 07:37:40.
This issue was reported by: `knut3`.
knut3 said at 2015-04-21 07:37:40
>Hi,
> 
>I wrote a tool and put it into site_cons/site_tools/my_tool.py. 
>(http://scons.org/wiki/ToolsForFools)
>According to the user guide "18.7. Where To Put Your Custom Builders and Tools"
>there have to be two functions: exists() and generate().
> 
>I use my tool and it works fine thus generate() was called. But the exists()
>function is never called. I put a "return False" there but no complains while
>running scons...
> 
>knut

**Comment by nobodyinperson** [comment]
Wondering about this as well. Is one supposed to call `exists()` inside `generate()` manually?

[The documentation](https://scons.org/doc/production/HTML/scons-man.html) really only mentions this:

> A tool specification must include two functions: generate(env, **kw) and exists(env). The generate function modifies the environment referenced by env to set up variables so that the tool can be executed; it may use any keyword arguments that the user supplies in kw (see below) to vary its initialization. **The exists function should return a true value if the tool is available**. 

What is the use of this `exists()` tool function, if it is not called upon `Environment` creation?

**Comment by mwichmann** [comment]
It's called when SCons initializes the tools, in a stanza that looks like this:
```
        for t in self.tools:
            tool = SCons.Tool.Tool(t)
            if tool.exists(env):
                env.Tool(tool)
```
It's not nearly as useful as it could be... if `exists` finds the program via a search, it knows where it is. But it can't return that path, because nobody is listening for that; so `generate` usually ends up searching for it again (or even, sigh, directly calling `exists` again).  Some of the existing tools just return `True` from `exists` without even checking....

**Comment by bdbaddog** [comment]
@nobodyinperson - ```exists()``` is called by SCons when it loads the default tools for the platform.  If you create your own tool, and then call ```env.Tool('mytool')```  exists() is not called.

See:

- https://github.com/SCons/scons/blob/master/src/engine/SCons/Tool/__init__.py#L1134
- https://github.com/SCons/scons/blob/master/src/engine/SCons/Tool/__init__.py#L1142
- https://github.com/SCons/scons/blob/master/src/engine/SCons/Tool/__init__.py#L1149

I'd say the current implementation should probably call ```exists()``` when ```env.Tool('mytool')```, but as yet it doesn't.  (Thus this bug)

So to sum up: Please do create a ```def exists()``` in your tool.  ```env.Tool()``` does already reference it, but never calls it. When this bug is resolved your tool will fail..

**Comment by mwichmann** [comment]
This behavior got some doc updates in #3778, which does not resolve this issue.
